agents/agents/guard/fraud.go

```diff
@@ -0,0 +1,792 @@
+package guard
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	ethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/synapsecns/sanguine/agents/types"
+	"github.com/synapsecns/sanguine/core/retry"
+)
+
+// handleSnapshot checks a snapshot for invalid states.
+// If an invalid state is found, initiate slashing and submit a state report.
+//
+//nolint:cyclop,gocognit
+func (g Guard) handleSnapshot(ctx context.Context, log ethTypes.Log) error {
+	fraudSnapshot, err := g.inboxParser.ParseSnapshotAccepted(log)
+	if err != nil {
+		return fmt.Errorf("could not parse snapshot accepted: %w", err)
+	}
+
+	// Verify each state in the snapshot.
+	for si, s := range fraudSnapshot.Snapshot.States() {
+		stateIndex, state := si, s
+		isSlashable, err := g.isStateSlashable(ctx, state)
+		if err != nil {
+			return fmt.Errorf("could not handle state: %w", err)
+		}
+		if !isSlashable {
+			continue
+		}
+
+		// Initiate slashing on origin.
+		_, err = g.txSubmitter.SubmitTransaction(ctx, big.NewInt(int64(state.Origin())), func(transactor *bind.TransactOpts) (tx *ethTypes.Transaction, err error) {
+			tx, err = g.domains[state.Origin()].LightInbox().VerifyStateWithSnapshot(
+				transactor,
+				int64(stateIndex),
+				fraudSnapshot.Payload,
+				fraudSnapshot.Signature,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("could not verify state with snapshot: %w", err)
+			}
+
+			return
+		})
+		if err != nil {
+			return fmt.Errorf("could not submit VerifyStateWithSnapshot tx: %w", err)
+		}
+
+		// Check if we should submit the state report.
+		shouldSubmit, err := g.shouldSubmitStateReport(ctx, fraudSnapshot)
+		if err != nil {
+			return fmt.Errorf("could not check if should submit state report: %w", err)
+		}
+		if !shouldSubmit {
+			return nil
+		}
+
+		// Submit the state report to summit.
+		srSignature, _, _, err := state.SignState(ctx, g.bondedSigner)
+		if err != nil {
+			return fmt.Errorf("could not sign state: %w", err)
+		}
+		ok, err := g.prepareStateReport(ctx, fraudSnapshot.Agent, g.summitDomainID)
+		if err != nil {
+			return fmt.Errorf("could not prepare state report on summit: %w", err)
+		}
+		if !ok {
+			continue
+		}
+		_, err = g.txSubmitter.SubmitTransaction(ctx, big.NewInt(int64(g.summitDomainID)), func(transactor *bind.TransactOpts) (tx *ethTypes.Transaction, err error) {
+			tx, err = g.domains[g.summitDomainID].Inbox().SubmitStateReportWithSnapshot(
+				transactor,
+				int64(stateIndex),
+				srSignature,
+				fraudSnapshot.Payload,
+				fraudSnapshot.Signature,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("could not submit state report with snapshot to summit: %w", err)
+			}
+
+			return
+		})
+		if err != nil {
+			return fmt.Errorf("could not submit SubmitStateReportWithSnapshot tx: %w", err)
+		}
```

Comment:
Might be overthinking this, but could we explore the generalization of submission of the report transaction in the future?

Something like "here's the chain Id, here's the signed report (of any type), here's some data to accompany it, please submit the report tx for me".

Reply:
Outlining this, as the core logic is pretty much the same:

https://github.com/synapsecns/sanguine/blob/f5eb4de20c67c5a570cdab50139334a51197e5e3/agents/agents/guard/fraud.go#L101-L117

https://github.com/synapsecns/sanguine/blob/f5eb4de20c67c5a570cdab50139334a51197e5e3/agents/agents/guard/fraud.go#L270-L287

etc